File: dat-rust/src/manager.rs
use crate::certificate::DatCertificate;
use crate::dat::Dat;
use crate::error::DatError;
use crate::payload::DatPayload;
use crate::util::{encode_base64_url_out, now_unix_timestamp, to_hex_u64_out};
use itertools::Itertools;
use std::sync::{PoisonError, RwLock, RwLockReadGuard, RwLockWriteGuard};

struct DatManagerState {
    issuer: Option<DatCertificate>,
    certificates: Vec<DatCertificate>,
}

pub struct DatManager {
    state: RwLock<DatManagerState>,
}

impl Default for DatManager {
    fn default() -> Self {
        Self::new()
    }
}

#[inline]
fn poisoned<T>(_: PoisonError<T>) -> DatError {
    DatError::InternalUnknown("dat manager lock is poisoned")
}

fn no_issuable_cause(certificates: &[DatCertificate]) -> DatError {
    let now = now_unix_timestamp();
    let (mut signable_seen, mut not_yet, mut ended) = (false, false, false);

    for certificate in certificates {
        if !certificate.signable() {
            continue;
        }
        signable_seen = true;
        if now < certificate.dat_issuance_start_seconds {
            not_yet = true;
        } else if now > certificate.dat_issuance_end_seconds {
            ended = true;
        }
    }

    if !signable_seen {
        DatError::CertVerifyOnly
    } else if not_yet {
        DatError::CertNotYetIssuable
    } else if ended {
        DatError::CertIssuanceEnded
    } else {
        DatError::CertExpired
    }
}

impl DatManager {
    pub fn new() -> Self {
        DatManager {
            state: RwLock::new(DatManagerState {
                issuer: None,
                certificates: vec![],
            }),
        }
    }

    #[inline]
    fn read(&self) -> Result<RwLockReadGuard<'_, DatManagerState>, DatError> {
        self.state.read().map_err(poisoned)
    }

    #[inline]
    fn write(&self) -> Result<RwLockWriteGuard<'_, DatManagerState>, DatError> {
        self.state.write().map_err(poisoned)
    }

    pub fn issue(&self, plain: &str, secure: &str) -> Result<String, DatError> {
        let state = self.read()?;
        if let Some(certificate) = state.issuer.as_ref() {
            Self::_issue(certificate, plain, secure)
        } else if state.certificates.is_empty() {
            Err(DatError::ManagerNoCertificate)
        } else {
            Err(DatError::ManagerNoIssuableCertificate(Box::new(
                no_issuable_cause(&state.certificates),
            )))
        }
    }

    pub fn parse<E: Into<DatError>>(
        &self,
        dat: impl TryInto<Dat, Error = E>,
    ) -> Result<DatPayload, DatError> {
        let dat = dat.try_into().map_err(Into::into)?;
        let cid = dat.cid;
        let state = self.read()?;
        if let Some(certificate) = state.certificates.iter().find(|x| x.cid == cid) {
            Self::_parse(certificate, dat)
        } else {
            Err(DatError::CertNotFound)
        }
    }

    pub fn parse_without_verify<E: Into<DatError>>(
        &self,
        dat: impl TryInto<Dat, Error = E>,
    ) -> Result<DatPayload, DatError> {
        let dat = dat.try_into().map_err(Into::into)?;
        let cid = dat.cid;
        let state = self.read()?;
        if let Some(certificate) = state.certificates.iter().find(|x| x.cid == cid) {
            Self::_parse_without_verify(certificate, dat)
        } else {
            Err(DatError::CertNotFound)
        }
    }

    pub fn export_cids(&self) -> Vec<u64> {
        self.state
            .read()
            .unwrap_or_else(PoisonError::into_inner)
            .certificates
            .iter()
            .map(|key| key.cid)
            .collect()
    }

    pub fn export(&self, verify_only: bool) -> Result<String, DatError> {
        let export = self
            .read()?
            .certificates
            .iter()
            .map(|key| key.export(verify_only))
            .collect::<Result<Vec<String>, DatError>>()?
            .join("\n");
        Ok(export)
    }

    pub fn export_certificates(&self) -> Result<Vec<DatCertificate>, DatError> {
        self.read()?
            .certificates
            .iter()
            .map(|x| x.try_clone())
            .collect()
    }

    pub fn import(&self, format: &str, clear: bool) -> Result<usize, DatError> {
        let format = format.trim();
        if format.is_empty() {
            return Ok(0);
        }
        let new_certificates = format
            .lines()
            .map(|s| s.parse::<DatCertificate>())
            .collect::<Result<Vec<DatCertificate>, DatError>>()?;
        self.import_certificates(new_certificates, clear)
    }

    pub fn import_certificates(
        &self,
        new_certificates: Vec<DatCertificate>,
        clear: bool,
    ) -> Result<usize, DatError> {
        if new_certificates.is_empty() {
            return Ok(0);
        }

        let mut apply_certs: usize = 0;
        let mut ids = new_certificates.iter().map(|x| x.cid).collect::<Vec<u64>>();
        ids.sort();
        ids.dedup();
        if ids.len() != new_certificates.len() {
            return Err(DatError::CertDuplicateCid);
        }

        let mut state = self.write()?;

        let mut certificates = if clear {
            vec![]
        } else {
            state
                .certificates
                .iter()
                .map(|x| x.try_clone())
                .collect::<Result<Vec<DatCertificate>, DatError>>()?
        };

        for certificate in new_certificates {
            if !certificates.contains(&certificate) {
                certificates.push(certificate);
                apply_certs += 1;
            }
        }

        let certificates = certificates
            .into_iter()
            .filter(|certificate| !certificate.expired())
            .sorted_by(|a, b| a.dat_issuance_end_seconds.cmp(&b.dat_issuance_end_seconds))
            .collect::<Vec<DatCertificate>>();

        let issuer = certificates
            .iter()
            .rev()
            .find(|x| x.issuable())
            .map(|x| x.try_clone())
            .transpose()?;

        state.issuer = issuer;
        state.certificates = certificates;

        Ok(apply_certs)
    }

    pub fn _issue<U: AsRef<[u8]>>(
        certificate: &DatCertificate,
        plain: U,
        secure: U,
    ) -> Result<String, DatError> {
        let mut ib = itoa::Buffer::new();
        let expire = now_unix_timestamp()
            .checked_add(certificate.dat_ttl_seconds)
            .ok_or(DatError::InternalUnknown(
                "now + dat_ttl_seconds overflowed u64",
            ))?;
        let expire = ib.format(expire);
        let plain = plain.as_ref();
        let secure = secure.as_ref();

        let mut v: String =
            String::with_capacity(60 + expire.len() + ((plain.len() + secure.len() + 100) * 4 / 3));
        let sk = &certificate.signature;

        v.push_str(expire);
        v.push('.');
        to_hex_u64_out(certificate.cid, &mut v);
        v.push('.');

        encode_base64_url_out(plain, &mut v);

        v.push('.');
        encode_base64_url_out(certificate.crypto.encrypt(secure)?, &mut v);

        let signature = sk.sign(v.as_bytes())?;
        v.push('.');
        encode_base64_url_out(&*signature, &mut v);
        Ok(v)
    }

    pub fn _parse(certificate: &DatCertificate, dat: Dat) -> Result<DatPayload, DatError> {
        certificate
            .signature
            .verify(dat.body_bytes(), &dat.signature)?;
        Self::_parse_without_verify(certificate, dat)
    }
    pub fn _parse_without_verify(
        certificate: &DatCertificate,
        dat: Dat,
    ) -> Result<DatPayload, DatError> {
        let plain = dat.plain()?;
        let secure = certificate.crypto.decrypt(dat.secure()?)?;

        Ok(DatPayload { plain, secure })
    }
}

File: dat-go/v4/manager.go
package dat

import (
	"slices"
	"strconv"
	"strings"
	"sync"
)

type Manager struct {
	issuer       *Certificate
	certificates []*Certificate
	certIndex    map[uint64]*Certificate
	mu           sync.RWMutex
}

func NewManager() *Manager {
	return &Manager{
		certificates: []*Certificate{},
		certIndex:    map[uint64]*Certificate{},
	}
}

func noIssuableCause(certificates []*Certificate) error {
	now := NowUnixTimestamp()
	signableSeen, notYet, ended := false, false, false

	for _, cert := range certificates {
		if !cert.Signable() {
			continue
		}
		signableSeen = true
		if now < cert.DatIssuanceStartSeconds {
			notYet = true
		} else if now > cert.DatIssuanceEndSeconds {
			ended = true
		}
	}

	switch {
	case !signableSeen:
		return ErrCertVerifyOnly
	case notYet:

		return ErrCertNotYetIssuable
	case ended:
		return ErrCertIssuanceEnded
	default:
		return ErrCertExpired
	}
}

func (m *Manager) Issue(plain, secure string) (string, error) {
	m.mu.RLock()
	issuer := m.issuer
	certificates := m.certificates
	m.mu.RUnlock()

	if issuer == nil {
		if len(certificates) == 0 {
			return "", ErrManagerNoCertificate
		}
		return "", ErrManagerNoIssuableCertificate.Wrap(noIssuableCause(certificates))
	}
	return m.IssueWithCertificate(issuer, plain, secure)
}

func (m *Manager) Parse(datStr string) (Payload, error) {
	d, err := ParseDat(datStr)
	if err != nil {
		return Payload{}, err
	}
	return m.ParseDat(d)
}

func (m *Manager) ParseDat(dat *Dat) (Payload, error) {
	m.mu.RLock()
	cert := m.certIndex[dat.Cid]
	m.mu.RUnlock()

	if cert == nil {
		return Payload{}, ErrCertNotFound
	}
	return m.ParseWithCertificate(cert, dat)
}

func (m *Manager) ParseWithoutVerify(datStr string) (Payload, error) {
	d, err := ParseDat(datStr)
	if err != nil {
		return Payload{}, err
	}
	return m.ParseDatWithoutVerify(d)
}

func (m *Manager) ParseDatWithoutVerify(dat *Dat) (Payload, error) {
	m.mu.RLock()
	cert := m.certIndex[dat.Cid]
	m.mu.RUnlock()

	if cert == nil {
		return Payload{}, ErrCertNotFound
	}
	return m.ParseWithoutVerifyWithCertificate(cert, dat)
}

func (m *Manager) ExportCids() []uint64 {
	m.mu.RLock()
	defer m.mu.RUnlock()

	cids := make([]uint64, len(m.certificates))
	for i, cert := range m.certificates {
		cids[i] = cert.Cid
	}
	return cids
}

func (m *Manager) Export(verifyOnly bool) string {
	m.mu.RLock()
	defer m.mu.RUnlock()

	var sb strings.Builder
	for i, cert := range m.certificates {
		if i > 0 {
			sb.WriteString("\n")
		}
		exported, _ := cert.Export(verifyOnly)
		sb.WriteString(exported)
	}
	return sb.String()
}

func (m *Manager) ExportCertificates() ([]*Certificate, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	certificates := make([]*Certificate, 0, len(m.certificates))
	for _, cert := range m.certificates {
		cloned, err := cert.TryClone()
		if err != nil {
			return nil, err
		}
		certificates = append(certificates, cloned)
	}
	return certificates, nil
}

func (m *Manager) Import(format string, clear bool) (int, error) {
	lines := strings.Split(format, "\n")
	var newCerts []*Certificate
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		cert, err := ParseCertificate(line)
		if err != nil {
			return 0, err
		}
		newCerts = append(newCerts, cert)
	}
	return m.ImportCertificates(newCerts, clear)
}

func (m *Manager) ImportCertificates(newCertificates []*Certificate, clear bool) (int, error) {
	ids := make(map[uint64]bool)
	for _, cert := range newCertificates {
		if ids[cert.Cid] {
			return 0, ErrCertDuplicateCid
		}
		ids[cert.Cid] = true
	}

	var renewCount int = 0

	m.mu.Lock()
	defer m.mu.Unlock()

	var certificates []*Certificate
	if clear {
		certificates = []*Certificate{}
	} else {
		certificates = slices.Clone(m.certificates)
	}

	for _, newCert := range newCertificates {
		found := false
		for _, cert := range certificates {
			if cert.Cid == newCert.Cid {
				found = true
				break
			}
		}
		if !found {
			certificates = append(certificates, newCert)
			renewCount++
		}
	}

	var filtered []*Certificate
	for _, cert := range certificates {
		if !cert.Expired() {
			filtered = append(filtered, cert)
		}
	}

	slices.SortFunc(filtered, func(a, b *Certificate) int {
		if a.DatIssuanceEndSeconds < b.DatIssuanceEndSeconds {
			return -1
		} else if a.DatIssuanceEndSeconds > b.DatIssuanceEndSeconds {
			return 1
		}
		return 0
	})

	var issuer *Certificate
	for i := len(filtered) - 1; i >= 0; i-- {
		if filtered[i].Issuable() {
			issuer = filtered[i]
			break
		}
	}

	certIndex := make(map[uint64]*Certificate, len(filtered))
	for _, cert := range filtered {
		certIndex[cert.Cid] = cert
	}

	m.issuer = issuer
	m.certificates = filtered
	m.certIndex = certIndex

	return renewCount, nil
}

func (m *Manager) IssueWithCertificate(certificate *Certificate, plain, secure string) (string, error) {
	encrypted, err := certificate.CryptoKey.Encrypt([]byte(secure))
	if err != nil {
		return "", err
	}

	buf := make([]byte, 0, 20+len(certificate.cidPreCopy)+
		base64URL.EncodedLen(len(plain))+1+base64URL.EncodedLen(len(encrypted))+1+180)
	buf = strconv.AppendUint(buf, NowUnixTimestamp()+certificate.DatTtlSeconds, 10)
	buf = append(buf, certificate.cidPreCopy...)
	buf = base64URL.AppendEncode(buf, []byte(plain))
	buf = append(buf, '.')
	buf = base64URL.AppendEncode(buf, encrypted)

	signature, err := certificate.SignatureKey.Sign(buf)
	if err != nil {
		return "", err
	}
	buf = append(buf, '.')
	buf = base64URL.AppendEncode(buf, signature)

	return string(buf), nil
}

func (m *Manager) ParseWithCertificate(certificate *Certificate, dat *Dat) (Payload, error) {
	if err := certificate.SignatureKey.Verify(dat.BodyBytes(), dat.Signature); err != nil {
		return Payload{}, err
	}
	return m.ParseWithoutVerifyWithCertificate(certificate, dat)
}

func (m *Manager) ParseWithoutVerifyWithCertificate(certificate *Certificate, dat *Dat) (Payload, error) {
	plain, err := dat.Plain()
	if err != nil {
		return Payload{}, err
	}
	secureEncoded, err := dat.Secure()
	if err != nil {
		return Payload{}, err
	}
	secure, err := certificate.CryptoKey.Decrypt(secureEncoded)
	if err != nil {
		return Payload{}, err
	}

	return Payload{
		Plain:  plain,
		Secure: secure,
	}, nil
}
